src/core/fetchers/segment/prioritizer.ts

```diff
@@ -136,87 +223,236 @@ export default class ObservablePrioritizer<T> {
+      return;
+    }
 
-    if (this._pendingPriority == null || this._pendingPriority >= priority) {
-      this._queue.splice(index, 1);
-      queueElement.trigger.next();
-      queueElement.trigger.complete();
+    const task = this._pendingTasks[pendingTasksIndex];
+    if (task.priority === priority) {
+      return;
     }
-  }
 
-  private _startObservable(obs : Observable<T>) : Observable<T> {
-    const onObservableFinish = () : void => {
-      this._numberOfPendingObservables--;
+    const oldPriority = task.priority;
+    task.priority = priority;
 
-      if (this._numberOfPendingObservables > 0) {
-        // still waiting for Observables to finish
-        return;
+    if (priority < oldPriority) {
+      if (this._isHighPriority(task)) {
+        this._interruptCancellableTasks();
       }
+      return;
+    }
 
-      this._pendingPriority = null;
+    if (this._minPendingPriority !== null &&
+        this._minPendingPriority <= this._prioritySteps.high)
+    {
+      // We could need to interrupt this task
+      this._interruptPendingTask(pendingTasksIndex);
+    }
+  }
 
-      if (this._queue.length === 0) {
-        return;
+  /**
+   * Interrupt and move back to the waiting queue all pending tasks that are
+   * low priority (having a higher priority number than
+   * `this._prioritySteps.low`).
+   */
+  private _interruptCancellableTasks() : void {
+    for (let i = 0; i < this._pendingTasks.length; i++) {
+      const pendingObj = this._pendingTasks[i];
+      if (pendingObj.priority >= this._prioritySteps.low) {
+        this._interruptPendingTask(i);
+
+        // The previous call could have a lot of potential side-effects.
+        // It is safer to re-start the function to not miss any pending
+        // task that needs to be cancelled.
+        return this._interruptCancellableTasks();
       }
+    }
+  }
 
-      this._pendingPriority = this._queue
-        .reduce((acc : number | null, elt) => {
-          return acc == null || acc > elt.priority ? elt.priority :
-                                                     acc;
-        }, null);
-
-      for (let i = 0; i < this._queue.length; i++) {
-        const elt = this._queue[i];
-        if (elt.priority === this._pendingPriority) {
-          this._queue.splice(i, 1);
-          i--;
-          elt.trigger.next();
-          elt.trigger.complete();
-        }
+  /**
+   * Start task which is at the given index in the waiting queue.
+   * The task will be removed from the waiting queue in the process.
+   * @param {number} index
+   */
+  private _startTask(index : number) : void {
+    const task = this._waitingQueue.splice(index, 1)[0];
+    task.trigger.next(true);
+  }
+
+  /**
+   * Move back pending task - which is at the given index in the _pendingTasks
+   * array - back to the waiting queue and interrupt it.
+   * @param {number} index
+   */
+  private _interruptPendingTask(index : number) : void {
```

Comment:
There is no chance that the pending task array could have changed between when we get the index, and when we call this function ?

I don't think so, but just asking

Reply:
Yes, it is something that is a little touchy. The caller has to be extra careful there.

I though this was not much different than calling something like "splice" directly as here too we explicitely ask for the index in the `_pendingTask` array, so it could be safer to just gives the corresponding task object.

I have mixed feelings about that.

Under one hand it would be the safer measure, under the other hand it just add more code in the callee - not only the loop but we also have to handle the case where the index is not found - to do the same thing.

What do you think we should do here?

Reply:
I tend to think that we may call the function with the task as argument. Moreover, it makes more sense by reading the code.
If the code surplus is not excessive, i would rather do that.

Reply:
Ok, will do

Reply:
Done